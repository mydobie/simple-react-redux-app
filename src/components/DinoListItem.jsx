import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-bootstrap';

const DinoListItem = (props) => {
  const { dinoName, dinoId, checked, changeCheckBox } = props;
  return (
    <li>
      {/* EXAMPLE: Checkbox with label */}
      <Form.Check
        onClick={(e) => {
          changeCheckBox(dinoId, e.target.checked);
        }}
        id={`cb_${dinoId}`}
        defaultChecked={checked}
        label={dinoName}
      />
    </li>
  );
};

DinoListItem.propTypes = {
  dinoName: PropTypes.string,
  dinoId: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  changeCheckBox: PropTypes.func,
};
DinoListItem.defaultProps = {
  dinoName: '',
  checked: false,
  changeCheckBox: () => {},
};

export default DinoListItem;
