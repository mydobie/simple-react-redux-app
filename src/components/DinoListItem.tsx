import React, { ReactElement } from 'react';
import { Form } from 'react-bootstrap';

// ** Main component type */
type DinoListItemProps = {
  dinoName: string;
  dinoId: string;
  checked: boolean;
  // EXAMPLE: Passing a function as a prop (so data bubbles to parent)
  changeCheckBox: (dinoId: string, isChecked: boolean) => void;
  testid?: string;
};

// *** Main component ***
const DinoListItem = ({
  dinoName,
  dinoId,
  checked,
  changeCheckBox,
  testid,
}: DinoListItemProps): ReactElement => (
  <li data-testid={testid}>
    {/* EXAMPLE: Checkbox with label */}
    <Form.Check
      onClick={(e) => {
        changeCheckBox(dinoId, (e.target as HTMLInputElement).checked);
      }}
      id={`cb_${dinoId}`}
      defaultChecked={checked}
      label={dinoName}
    />
  </li>
);

export default DinoListItem;
