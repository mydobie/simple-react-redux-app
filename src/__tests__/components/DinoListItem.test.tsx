/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable react/react-in-jsx-scope */

// @ts-ignore
import { render, fireEvent } from '@testing-library/react';
import { axe } from 'jest-axe';
import DinoListItem from '../../components/DinoListItem';

// EXAMPLE: Mocking a function sent in as a prop
const mockChangeCheckbox = jest.fn((/* id, checked */) => {});

describe('Loading icon tests', () => {
  afterEach(() => {
    mockChangeCheckbox.mockClear();
  });

  let container: HTMLElement;

  beforeEach(() => {
    container = render(
      <DinoListItem
        dinoName='tRex'
        dinoId='4'
        checked={false}
        changeCheckBox={mockChangeCheckbox}
      />
    ).container;
  });

  test('Is accessible', async () => {
    const results = await axe(`<main><ul>${container.innerHTML}</ul></main>`); // NOTE ul is required to prevent landmark error
    expect(results).toHaveNoViolations();
  });

  test('When checkbox is clicked, changeCheckBox is called', () => {
    // EXAMPLE: Checking for a React Bootstrap element.
    // In the component it is listed as Form.Check
    // If you are unsure what an element is really called, then look at the snapshot
    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelectorAll('input[type="checkbox"]')[0]);

    expect(mockChangeCheckbox).toHaveBeenCalledTimes(1);
    expect((mockChangeCheckbox.mock.calls[0] as string[])[0]).toBe('4');
    expect((mockChangeCheckbox.mock.calls[0] as boolean[])[1]).toBe(true); // second attribute sent
  });
});
