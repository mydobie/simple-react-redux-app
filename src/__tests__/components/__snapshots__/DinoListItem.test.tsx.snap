// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading icon tests When checkbox is clicked, changeCheckBox is called 1`] = `
<div>
  <li>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="cb_4"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="cb_4"
        title=""
      >
        tRex
      </label>
    </div>
  </li>
</div>
`;
